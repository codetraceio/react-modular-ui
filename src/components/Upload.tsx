import React, { DragEvent, PropsWithChildren, useRef, useState, useCallback, useContext } from "react";

import { className } from "../utils/className";
import { ThemeContext } from "./ThemeContext";

export interface UploadProps {
  size?: string | number;
  color?: string;
  disabled?: boolean;
  checked?: boolean;
  name?: string;
  theme?: string;

  onChange?: (files: FileList) => void
}

export default function Upload(props: PropsWithChildren<UploadProps>) {
  const { onChange } = props;
  const theme = useContext(ThemeContext);
  const inputRef = useRef<HTMLInputElement>(null);

  const [active, setActive] = useState(false);

  const handleDragOver = useCallback((event: DragEvent<HTMLDivElement>) => {
    event.preventDefault();
    setActive(true);
  }, [setActive]);

  const handleDragLeave = useCallback(() => {
    setActive(false);
  }, []);

  const handleDrop = useCallback((event: DragEvent<HTMLDivElement>) => {
    event.stopPropagation();
    event.preventDefault();

    const files = event.dataTransfer.files;

    setActive(false);

    if (typeof onChange === "function") {
      onChange(files);
    }
  }, [setActive, onChange]);

  const handleClick = useCallback(() => {
    if (!inputRef.current) {
      return;
    }

    inputRef.current.click();
  }, []);

  const handleChange = useCallback(() => {
    const files = inputRef.current.files;

    if (typeof onChange === "function") {
      onChange(files);
    }
  }, [onChange]);

  return (
    <div
      className={className("upload")}
      data-name={props.name}
      data-active={active}
      data-theme={props.theme ?? theme}
      onDragOver={handleDragOver}
      onDragLeave={handleDragLeave}
      onDrop={handleDrop}
      onClick={handleClick}
    >
      <input
        type="file"
        name={props.name}
        style={{display: "none"}}
        ref={inputRef}
        onChange={handleChange}
      />
      <div>
        {props.children}
      </div>
    </div>
  );
}
