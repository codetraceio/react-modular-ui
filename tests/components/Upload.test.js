import * as React from "react";
import * as renderer from "react-test-renderer";

import Upload from "../../src/components/Upload";

describe("Upload", () => {
  it("should render Upload with default text", () => {
    const tree = renderer.create(
      <Upload
        text="Something"
        successText="Something else"
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render Upload with success text when upload succeed", () => {
    const tree = renderer.create(
      <Upload
        text="Something"
        successText="Something else"
        success
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
