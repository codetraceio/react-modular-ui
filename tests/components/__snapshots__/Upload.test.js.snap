// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload should render Upload with default text 1`] = `
<div
  className="c-upload"
  onClick={[Function]}
>
  <input
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div />
</div>
`;

exports[`Upload should render Upload with success text when upload succeed 1`] = `
<div
  className="c-upload"
  onClick={[Function]}
>
  <input
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div />
</div>
`;
